rules/TypeDeclaration/Rector/ClassMethod/NarrowObjectReturnTypeRector.php

```diff
@@ -0,0 +1,264 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\TypeDeclaration\Rector\ClassMethod;
+
+use PhpParser\Node;
+use PhpParser\Node\Identifier;
+use PhpParser\Node\Name\FullyQualified;
+use PhpParser\Node\Stmt\ClassMethod;
+use PHPStan\Reflection\ClassReflection;
+use PHPStan\Type\ObjectType;
+use Rector\PhpParser\Node\BetterNodeFinder;
+use Rector\Rector\AbstractRector;
+use Rector\Reflection\ReflectionResolver;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * Narrows return types from generic object or parent class to specific class.
+ *
+ * @see \Rector\Tests\TypeDeclaration\Rector\ClassMethod\NarrowObjectReturnTypeRector\NarrowObjectReturnTypeRectorTest
+ */
+final class NarrowObjectReturnTypeRector extends AbstractRector
+{
+    public function __construct(
+        private readonly BetterNodeFinder $betterNodeFinder,
+        private readonly ReflectionResolver $reflectionResolver,
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Narrows return type from generic object or parent class to specific class',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+final class TalkFactory extends AbstractFactory
+{
+    protected function build(): object
+    {
+        return new ConferenceTalk();
+    }
+}
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+final class TalkFactory extends AbstractFactory
+{
+    protected function build(): ConferenceTalk
+    {
+        return new ConferenceTalk();
+    }
+}
+CODE_SAMPLE
+                ),
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+final class TalkFactory
+{
+    public function createConferenceTalk(): Talk
+    {
+        return new ConferenceTalk();
+    }
+}
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+final class TalkFactory
+{
+    public function createConferenceTalk(): ConferenceTalk
+    {
+        return new ConferenceTalk();
+    }
+}
+CODE_SAMPLE
+                ),
+            ],
+        );
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [ClassMethod::class];
+    }
+
+    /**
+     * @param ClassMethod $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (! $node instanceof ClassMethod) {
+            return null;
+        }
+
+        $returnType = $node->returnType;
+
+        if (! $returnType instanceof Identifier && ! $returnType instanceof FullyQualified) {
+            return null;
+        }
+
+        $classReflection = $this->reflectionResolver->resolveClassReflection($node);
+        if (! $classReflection instanceof ClassReflection) {
+            return null;
+        }
+
+        if (! $classReflection->isFinal() && ! $node->isFinal()) {
```

Comment:
`classReflection->isFinal()` support `@final` docblock, should use `classReflection->isFinalByKeyword()` instead.